### ISSUE TITLE
Definitive solution for ChildContext

### ISSUE THREAD

**Issue by jakebeal** [Protelis/Protelis#171]
I believe that BBN is still using the patch solution in #140 for dealing with `self` accesses from with an aligned map. Since that's disapproved, we need to figure out what the right way to deal with this is.

**Comment by jakebeal** [comment]
An example of  a current BBN implementation of ChildContext as required for successful execution within aligned-map: https://github.com/map-dcomp/map-code/blob/d4abe18feacec656ed86d3e1f591d6d068fa74de/src/MAP-Agent/src/main/java/com/bbn/map/Controller.java#L603

Is this following the recommended pattern, or doing something wrong?

**Comment by DanySK** [comment]
Hi Jake,
the recommended pattern is to return an instance of the same class or a subclass of the parent context. Otherwise, methods available at the "top level" `self` won't be available inside restricted contexts. I guess that in your example `Controller` is a subclass of `AbstractExecutionContext`. If so, I believe that invoking `self.debug(...)` would work outside of an `alignedMap`, and fail otherwise, as `ChildContext` does not have such method.
Now, if a function uses a `self.` call internally, that function won't be safe to use inside an `alignedMap`, making such construct even more subtle than it already is.

## Some thoughts

This is one of the parts of Protelis which is most critical, as we did not have enough grasp of device and networking modelling when we designed the language (at least, I didn't, I'm not sure that I do now, but sure I have many more insights). In #146 I suggested a configuration via dependency injection. It's a possibility for dealing with diverse computational contexts, but it is not applicable to the current Protelis structure.

I believe the driver force behind this request is related to the serialization of the contexts. Is this correct? If not, which is the use case that is not well (or at all) supported by the recommended implementation pattern?

We may try to experiment with the Strangely Recurring Template Pattern to emulate a self type in Java (I used it in the past), thus forcing the platform designer to stick to the pattern:
```java
public interface ExecutionContext<S extends ExecutionContext<? extends S>> {
    ...
   S instance();
   ...
}
```
This should guarantee that any method invokable on `self` always exists.

**Comment by DanySK** [comment]
Possible implementation with the generic recursion is available in #225.
The `ExecutionContext` should not be used as god class hosting most of the computation, should instead contain as less machinery as possible. With reference to the implementation in P2Protelis, the "child" context should be the only context, threading, ProtelisVM management, execution, and networking should not be part of it.

**Comment by jakebeal** [comment]
Sorted now, at least architecturally.
